Analytics

KarateCowboy opened the pull request.

jdx commented on src/analytics.js:
```diff
@@ -0,0 +1,122 @@
+    this.pluginVersion = pluginVersion
+    this.version = version
+    this.analyticsPath = "  "
+    this.os = OS.platform()
```
you could grab this from the config object I think

jdx commented on src/analytics.js:
```diff
@@ -0,0 +1,122 @@
+        'User-Agent': 3
+      }
+    }
+    await HTTPS.request(options)
```
you should use http-call like I do in the updater instead of directly using the https library. This won't even function since it doesn't output a promise.

jdx commented on src/analytics.test.js:
```diff
@@ -0,0 +1,141 @@
+        HTTP.request = jest.fn()
+        AnalyticsCommand.skipAnalytics = jest.fn(() => { return true })
+        await AnalyticsCommand.submitAnalytics()
+        expect(HTTP.request).not.toHaveBeenCalled()
```
I suggest using nock instead to validate http calls

ransombriggs replied:
@KarateCowboy I agree, could you use nock here?

ransombriggs commented on src/analytics.js:
```diff
@@ -0,0 +1,94 @@
+    this.version = this.config.config.version
+    this.platform = OS.platform()
+    this.arch = OS.arch()
+    this.analyticsPath = '~/.local/share/heroku/analytics.json'
```
@KarateCowboy please do not hard code this, you can get the cache directory from here

https://github.com/heroku/cli-engine-command/blob/master/src/config.js#L52

ransombriggs commented on src/analytics.js:
```diff
@@ -0,0 +1,94 @@
+  static async netrcLogin () {
+    if (process.env['HEROKU_API_KEY'] !== undefined || process.env['HEROKU_API_KEY'].length > 0) return false
+    let netrc = new Netrc()
+    return await netrc.machines['api.heroku.com'].login
```
@KarateCowboy please use [`apiHost`](https://github.com/heroku/cli-engine-command/blob/master/src/heroku.js#L19) from [`Vars`](https://github.com/heroku/cli-engine/blob/2476a1adcd2ddb6e610c69436508107b77c50c94/src/legacy.js#L6) to figure out which host to use.

ransombriggs commented on src/analytics.js:
```diff
@@ -0,0 +1,94 @@
+    return false
+  }
+
+  static analyticsPath(){
```
@KarateCowboy could you define or remove?

jdx commented on package.json:
```diff
@@ -15,13 +15,15 @@
     "levenshtein": "^1.0.5",
     "lodash.maxby": "4.6.0",
     "lodash.uniqby": "^4.7.0",
-    "log-chopper": "0.1.1",
```
can you leave these as exact dependencies?

jdx commented on package.json:
```diff
@@ -15,13 +15,15 @@
     "execa": "0.6.3",
     "find-node-modules": "^1.0.4",
     "fs-extra": "2.1.2",
+    "http-call": "1.1.0",
```
doesn't look like you're using this dependency

jdx commented on .gitignore:
```diff
@@ -1,4 +1,6 @@
 coverage
 lib
 node_modules
+.idea
```
this should go into your global gitignore

jdx commented on src/analytics.js:
```diff
@@ -0,0 +1,106 @@
+    this.command = command
+    this.plugin = plugin
+    this.pluginVersion = pluginVersion
+    this.platform = OS.platform()
```
should be available from `config.platform`

jdx commented on src/analytics.js:
```diff
@@ -0,0 +1,106 @@
+    this.arch = OS.arch()
+  }
+
+  get version (): string {
```
I might be missing something but this doesn't seem necessary since you could just populate the field in the constructor like you do for other properties

jdx commented on src/analytics.js:
```diff
@@ -0,0 +1,106 @@
+  }
+
+  async recordEnd () {
+    if (AnalyticsCommand.skipAnalytics() || process.argv.length < 2) {
```
seems like it would be good to pass in the arguments rather than grabbing them from process so you could write tests for this

KarateCowboy replied:
that was literally copied over verbatim from the Go code. I'm not even sure the logic around it.

jdx commented on src/analytics.js:
```diff
@@ -0,0 +1,106 @@
+    return this.config.version
+  }
+
+  async recordEnd () {
```
can this be named `record()` instead?

jdx commented on src/analytics.js:
```diff
@@ -0,0 +1,106 @@
+    return config.dirs.cache + '/analytics.json'
+  }
+
+  static async netrcLogin () {
```
why is this async?

jdx commented on src/analytics.js:
```diff
@@ -0,0 +1,106 @@
+  }
+
+  static async netrcLogin () {
+    // flow$ignore
```
is this because of the return false? it should probably return undefined instead then this should work I think

jdx commented on src/commands/plugins/update.test.js:
```diff
@@ -1,6 +1,9 @@
 import PluginsUpdate from './update'
+import fs from 'fs-extra'
+fs.readFileSync = jest.fn()
+fs.readFileSync.mockReturnValue(' { "skipAnalytics": true }')
```
seems like this is testing the implementation of the config module. I think it would be better to do `new Config({skipAnalytics: true})`

jdx reviewed (commented):
some initial feedback

KarateCowboy commented:
@dickeyxxx I addressed all of your comments except for the one about the process args, which I made a note about. Could you give some guidance in that case?

jdx commented on .gitignore:
```diff
@@ -1,5 +1,8 @@
 coverage
 lib
+node_modules
+.idea
```
this should go in your global gitignore

KarateCowboy replied:
fixed

jdx commented on .gitignore:
```diff
@@ -1,5 +1,8 @@
 /node_modules
 /test/links/test-foo/node_modules
 tmp
+
```
extra newline

KarateCowboy replied:
fixed

jdx commented on package.json:
```diff
@@ -20,8 +20,10 @@
     "log-chopper": "0.1.1",
-    "moment": "2.18.1",
-    "rwlockfile": "1.3.7",
+    "mermaid": "^7.0.0",
```
is this used for anything?

jdx commented on package.json:
```diff
@@ -20,8 +20,10 @@
+    "mermaid": "^7.0.0",
+    "moment": "2.17.1",
+    "netrc-parser": "^2.0.1",
+    "rwlockfile": "^1.3.6",
```
can you make these exact dependencies?

jdx commented on src/analytics.js:
```diff
@@ -0,0 +1,103 @@
+import fs from 'fs-extra'
+import { default as OS } from 'os'
+// flow$ignore
+import type Config from 'cli-engine-config'
```
what is being ignored here?

KarateCowboy replied:
```
src/analytics.js:5
  5: import type Config from 'cli-engine-config'
                 ^^^^^^ Default import from `cli-engine-config`. This module has no default export. Did you mean `import {Config} from "cli-engine-config"`?
```

I tried doing:
```
import { type Config } from 'cli-engine-config'
```
but then everything goes kaboom

jdx commented on src/analytics.js:
```diff
@@ -0,0 +1,103 @@
+    this.plugin = plugin
+    this.pluginVersion = pluginVersion
+    this.platform = config.platform
+    this.arch = OS.arch()
```
should be available as `config.arch` I think

jdx commented on src/analytics.js:
```diff
@@ -0,0 +1,103 @@
+    this.version = this.config.version
+  }
+
+  async record () {
```
is this ever called? also, does it need to be async?

jdx commented on src/analytics.js:
```diff
@@ -0,0 +1,103 @@
+  }
+
+  async record () {
+    if (AnalyticsCommand.skipAnalytics() || process.argv.length < 2) {
```
this skips recording if either the analytics need to be skipped or there is no command. This could probably be done inside of cli.js for v6

jdx commented on src/analytics.js:
```diff
@@ -0,0 +1,103 @@
+      return
+    }
+    let analyticsJSON
+    if (fs.existsSync(AnalyticsCommand.analyticsPath(this.config))) {
```
couldn't this be an instance property instead?

```js
fs.existsSync(this.analyticsPath)
```

KarateCowboy replied:
The static `submitAnalytics` also uses it.

jdx commented on src/analytics.js:
```diff
@@ -0,0 +1,103 @@
+    let analyticsJSON
+    if (fs.existsSync(AnalyticsCommand.analyticsPath(this.config))) {
+      let analyticsData = fs.readFileSync(AnalyticsCommand.analyticsPath(this.config), 'utf8')
+      analyticsJSON = JSON.parse(analyticsData)
```
what happens if this fails?

jdx commented on src/analytics.js:
```diff
@@ -0,0 +1,103 @@
+      version: this.config.version,
+      platform: this.platform
+    })
+    fs.writeFileSync(AnalyticsCommand.analyticsPath(this.config), JSON.stringify(analyticsJSON), 'utf8')
```
you could use fs.writeJson instead here

KarateCowboy replied:
changed

jdx commented on src/analytics.js:
```diff
@@ -0,0 +1,103 @@
+    fs.writeFileSync(AnalyticsCommand.analyticsPath(this.config), JSON.stringify(analyticsJSON), 'utf8')
+  }
+
+  static async submitAnalytics (config: Config) {
```
why is this static but the record function is an instance method? seems like it would be better for them to both be instance methods

jdx commented on src/analytics.js:
```diff
@@ -0,0 +1,103 @@
+      host = 'https://cli-analytics.heroku.com/record'
+    }
+    let out = new Output(config)
+    let http = new HTTP(out) // flow$ignore
```
what is being ignored here?

jdx commented on src/analytics.js:
```diff
@@ -0,0 +1,103 @@
+      const parsedHost: Object = URI.parse(host)
+      host = parsedHost.protocol + `//` + parsedHost.hostname + parsedHost.path
+      if (parsedHost.port) {
+        // flow$ignore
```
what is this ignore for?

jdx commented:
depends on https://github.com/heroku/cli-engine-command/pull/17 for CLI_ENGINE_ANALYTICS_URL to work
